core/Enclave/enclave_hsm.cpp

```diff
@@ -43,767 +42,598 @@
-    uint32_t real_blob_len = sgx_calc_gcm_data_size(0, SGX_AES_KEY_SIZE);
-
-    if (real_blob_len == UINT32_MAX)
-        return SGX_ERROR_UNEXPECTED;
-
-    if (req_blob_size != NULL) {
-        *req_blob_size = real_blob_len;
-        return SGX_SUCCESS;
     }
-
-    if (cmk_blob == NULL || cmk_blob_size != real_blob_len)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    uint8_t* tmp = (uint8_t *)malloc(SGX_AES_KEY_SIZE);
-    if (tmp == NULL)
-        return SGX_ERROR_OUT_OF_MEMORY;
-
-    ret = sgx_read_rand(tmp, SGX_AES_KEY_SIZE);
-    if (ret != SGX_SUCCESS) {
-        free(tmp);
-        return ret;
+    switch (cmk->metadata.keyspec)
+    {
+    case EH_AES_GCM_128:
+    case EH_AES_GCM_192:
+    case EH_AES_GCM_256:
+        ret = ehsm_create_aes_key(cmk);
+        break;
+    case EH_RSA_2048:
+    case EH_RSA_3072:
+    case EH_RSA_4096:
+        ret = ehsm_create_rsa_key(cmk);
+        break;
+    case EH_EC_P224:
+    case EH_EC_P256:
+    case EH_EC_P384:
+    case EH_EC_P512:
+    case EH_SM2:
+        ret = ehsm_create_ec_key(cmk);
+        break;
+    case EH_SM4_CTR:
+    case EH_SM4_CBC:
+        ret = ehsm_create_sm4_key(cmk);
+        break;
+    default:
+        break;
     }
 
-    ret = sgx_gcm_encrypt(&g_domain_key, SGX_AES_KEY_SIZE, tmp, 0, NULL, cmk_blob_size, (sgx_aes_gcm_data_ex_t *)cmk_blob);
-
-    memset_s(tmp, SGX_AES_KEY_SIZE, 0, SGX_AES_KEY_SIZE);
-
-    free(tmp);
-
     return ret;
 }
 
-/*
- * struct cipherblob {
- *    OUT uint8_t ciphertext[SGX_AES_KEY_SIZE];
- *    OUT uint8_t iv[EH_AES_GCM_IV_SIZE];   // 12B
- *    OUT uint8_t mac[EH_AES_GCM_MAC_SIZE]; // 16B
- * }
- */
-sgx_status_t enclave_aes_encrypt(const uint8_t *aad, size_t aad_len,
-                             const uint8_t *cmk_blob, size_t cmk_blob_size,
-                             const uint8_t *plaintext, size_t plaintext_len,
-                             uint8_t *cipherblob, size_t cipherblob_len)
+sgx_status_t enclave_encrypt(const ehsm_keyblob_t *cmk, size_t cmk_len,
+                             const ehsm_data_t *aad, size_t aad_len,
+                             const ehsm_data_t *plaintext, size_t plaintext_len,
+                             ehsm_data_t *ciphertext, size_t ciphertext_len)
 {
     sgx_status_t ret = SGX_ERROR_UNEXPECTED;
-
-    if (cmk_blob == NULL)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    uint32_t real_cmk_blob_size = sgx_calc_gcm_data_size(0, SGX_AES_KEY_SIZE);
-    if (UINT32_MAX == real_cmk_blob_size || cmk_blob_size < real_cmk_blob_size)
-        return SGX_ERROR_INVALID_PARAMETER;
-
-    uint32_t enc_key_size = sgx_get_gcm_ciphertext_size((sgx_aes_gcm_data_ex_t *)cmk_blob);
-    if (enc_key_size == UINT32_MAX || enc_key_size != sizeof(sgx_key_128bit_t)) {
-        printf("enc_key_size:%d is not expected: %lu.\n", enc_key_size, sizeof(sgx_key_128bit_t));
+    if (cmk == NULL || cmk->metadata.origin != EH_INTERNAL_KEY || plaintext == NULL || ciphertext == NULL)
+    {
```

Comment:
same here, all the input's length need to be checked due to it's the entry to the ecall

Reply:
follow the line145 to check each buffer's length